### PR TITLE
Pass RobotModel to kinematics plugins

### PR DESCRIPTION
**Pull request by rhaschke**
This is an alternative implemenation of #150 and #1153 to directly pass a RobotModel (RM) to kinematics plugins, thus
1) saving some loading time by reusing the `RobotModel` (particularly if there are dozens of groups)
2) ensuring a consistent RM (particularly, joint limits) is used throughout the code

#150 eventually got stuck due to circular reference of `RobotModelPtr` as detailed in https://github.com/ros-planning/moveit/pull/150#issuecomment-433106660.
To avoid/break this circular reference #1153 transferred ownership of KinematicSolvers (KS) from JointModelGroup (JMG) to KinematicsPluginLoader (KPL) / RobotModelLoader (RML). While this works, I think this is an evil hack. Ownership definitely belongs to JMG. This already becomes obvious from the fact that a RM with JMGs can (theoretically) exist without a KPL.

In this PR, ownership stays with JMGs. To break the circular reference, the RM is passed as reference to the KS. There, in contrast to #150, a new shared_ptr is created from the raw pointer, thus performing its own use counting. This way we decouple the use of the RM within the external KS from MoveIt. MoveIt just ensures that the RM is alive as long as the KS is. The KinematicsBase destructor could even check that the decoupled shared_ptr was correctly released by KS (and issue a warning if not).

To allow a smooth migration path, KS implementations should be able to choose which API for `initialize()` to implement. This is not (yet) considered in #1153. Here, KPL first tries the new API and - on failure - the old API. KinematicsBase provides default implementations for them, returning false, i.e. indicating failure.

Adaptation of CachedIK was a little bit tricky: This one wraps an arbitrary other KS. To relay `initialize()` calls of the wrapper to the available implementation(s) of the wrapped KS, I use SFINAE magic. This works for KDL IK, which already has the old API removed.

**Please, keep commit history when merging!**

### CODE REVIEW

**Comment by rhaschke** [comment]
> A JMG however always needs to be feed with `setSolverAllocators` by someone else, so it is not really self contained in the first place.

That's true.

> This dependency could be made more explicit and then also for your case of having JMG/RM without KPL it would be obvious that you need a replacement class (and storage) for KPL.

However, this dependency can but doesn't need to be satisfied by KPL. In principle you could directly link against a KS plugin and pass `std::make_shared` as an allocator. 
Again, while the JMG _needs_ a KS to be operational (and for this reason should own it), the KPL just _caches_  the shared ptrs. This is a fundamental semantic difference.

> I think the right solution would actually be to have RobotState store (and accept in it's constructor) a raw pointer.

I disagree and I hope you don't consider that as viable option as well. Changing this at such a central spot, would dramatically increase fragility of the framework.

> As said in the inline comment, a shared pointer communicates (potential) transfer of ownership.

Agreed.

> There is however no intention to do transfer the ownership of RobotModel to some random plugin.

Indeed. Which is why I pass a reference, indicating that this is a borrowed pointer.

File: moveit_core/kinematics_base/src/kinematics_base.cpp [hunk 1]
```diff
@@ -37,42 +37,87 @@
...
+static void no_deleter(const moveit::core::RobotModel*)
 {
-  robot_description_ = robot_description;
+}
+
+void KinematicsBase::storeValues(const moveit::core::RobotModel& robot_model, const std::string& group_name,
+                                 const std::string& base_frame, const std::vector<std::string>& tip_frames,
+                                 double search_discretization)
+{
+  robot_model_ = moveit::core::RobotModelConstPtr(&robot_model, &no_deleter);
```
[the remark this answers is not in the document]
Comment (rhaschke) [reply moveit_core/kinematics_base/src/kinematics_base.cpp]: I wouldn't pull in RobotState into the discussion. Passing a reference instead of a shared_ptr into the KS plugin, I indicate that ownership _is not_ transferred. Rather, I guarantee that this reference is alive as long as the KS is. It's the decision of the KS plugin how to proceed with the reference. To comply with the existing RobotState API, I form a new shared_ptr, which can be freely used by all entities of the KS - as long as the KS is alive.
This is safe and clean.

File: moveit_core/kinematics_base/include/moveit/kinematics_base/kinematics_base.h [hunk 2]
```diff
@@ -380,21 +386,31 @@ class KinematicsBase
...
+   * @param group_name The group for which this solver is being configured
+   * @param base_frame The base frame in which all input poses are expected.
+   * This may (or may not) be the root frame of the chain that the solver operates on
+   * @param tip_frames The tip of the chain
+   * @param search_discretization The discretization of the search when the solver steps through the redundancy
+   * @return true if initialization was successful, false otherwise
+   *
+   * When returning false, the KinematicsPlugingLoader will use the old method, passing a robot_description.
+   * Default implementation returns false and issues a warning to implement this new API.
+   * TODO: Make this method virtual after some soaking time, replacing the fallback.
```
[the remark this answers is not in the document]
Comment (rhaschke) [reply moveit_core/kinematics_base/include/moveit/kinematics_base/kinematics_base.h]: Yes, that's what I meant ;-)

Comment (rhaschke) [inline moveit_core/kinematics_base/include/moveit/kinematics_base/kinematics_base.h]:
```suggestion
   * TODO: Make this method pure virtual after some soaking time, replacing the fallback.
```

File: moveit_core/kinematics_base/src/kinematics_base.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (simonschmeisser) [reply moveit_core/kinematics_base/src/kinematics_base.cpp]: I agree with "safe enough" but not with clean. As we are currently defining the API that is going to used for the entire lifetime of melodic I think we need to also consider the API of RobotState and look at the whole picture.

[the remark this answers is not in the document]
Comment (rhaschke) [reply moveit_core/kinematics_base/src/kinematics_base.cpp]: Can we continue this discussion on the phone? You can find my phone number from [here](https://ekvv.uni-bielefeld.de/pers_publ/publ/Home.jsp).

[the remark this answers is not in the document]
Comment (simonschmeisser) [reply moveit_core/kinematics_base/src/kinematics_base.cpp]: After the phone call I still have my reservations against that construct but since this approach is otherwise minimally invasive I'm fine with having it merged. Let's await reviews from others

[the remark this answers is not in the document]
Comment (rhaschke) [reply moveit_core/kinematics_base/src/kinematics_base.cpp]: I believe, this `no-deleter` pointer just feels weird because it's unusual. However it's a clean solution to decouple the usage counting in MoveIt (who actually owns the passed RobotModel) from the one in the external plugin lib (which should only receive and manage a borrowed reference). If the external plugin lib messes up it's usage counting (e.g. introducing circular references), this should not affect MoveIt.

**Comment by rhaschke** [comment]
One idea of this PR was to ensure that the RobotModels used across the system are indeed identical and not out-of-sync. Creating a copy would raise this issue again.
If somebody should ever change the RobotModel on the fly, he should also ensure to reload the solvers, shouldn't he?

**Comment by jonbinney** [comment]
Sounds like there's two somewhat separate issues here:
1. making sure that the initial robot model used by the kinematics matches what the planners are using, including any configuration done by parameters outside of the URDF
2. keeping the model used by the kinematics solver up-to-date if it changes elsewhere in the code

Keeping a copy would satisfy point 1. To satisfy point 2, some amount of reloading may be necessary, but I think between reloads it would make sense for the IK solver's internally stored robot model to match what its internal datastructures were constructed with.

**Comment by rhaschke** [comment]
While the second issue (updating the kinematics solvers or other parts of the framework if the RobotModel has changed) is an interesting and useful feature request, we should not tackle this in the scope of this PR. I think this feature will require more intense discussions on how to handle such changes. Please open a feature request for this.

**Comment by rhaschke** [comment]
@v4hn, @davetcoleman: Would be great to get feedback on the proposed approach. I we agree, it remains to tackle the other plugins (`lma`, `srv`, `ikfast`) in a similar fashion as I already did for `kdl`.

**Comment by rhaschke** [comment]
Rebased to resolve conflicts.

File: MIGRATION.md [hunk 3]
```diff
@@ -8,6 +8,7 @@ API changes in MoveIt! releases
 - The move_group capability ``ExecuteTrajectoryServiceCapability`` has been removed in favor of the improved ``ExecuteTrajectoryActionCapability`` capability. Since Indigo, both capabilities were supported. If you still load default capabilities in your ``config/launch/move_group.launch``, you can just remove them from the capabilities parameter. The correct default capabilities will be loaded automatically.
 - Deprecated method ``CurrentStateMonitor::waitForCurrentState(double wait_time)`` was finally removed.
 - Renamed ``RobotState::getCollisionBodyTransforms`` to ``getCollisionBodyTransform`` as it returns a single transform only.
+- KinematicsBase: Deprecate members tip_frame_, search_discretization_. Use tip_frames_ and redundant_joint_discretization_ instead.
```
Comment (davetcoleman) [inline MIGRATION.md]: This deprecation change could easily be merged in via a separate PR, making this one easier to review

File: moveit_core/kinematics_base/CMakeLists.txt [hunk 4]
```diff
@@ -2,6 +2,7 @@ set(MOVEIT_LIB_NAME moveit_kinematics_base)
 
 add_library(${MOVEIT_LIB_NAME} src/kinematics_base.cpp)
 set_target_properties(${MOVEIT_LIB_NAME} PROPERTIES VERSION ${${PROJECT_NAME}_VERSION})
+target_compile_options(${MOVEIT_LIB_NAME} PRIVATE -Wno-deprecated-declarations)
```
Comment (davetcoleman) [inline moveit_core/kinematics_base/CMakeLists.txt]: what is the reasoning for this? can we keep this warning on? if not, add comment as to why not

File: moveit_core/kinematics_base/include/moveit/kinematics_base/kinematics_base.h [hunk 5]
```diff
@@ -569,31 +578,22 @@ class KinematicsBase
...
-                           // previously generated custom ik solvers like IKFast
 
-  double search_discretization_;  // DEPRECATED - this variable only still exists for backwards compatibility
-                                  // with previous implementations.  Discretization values for each joint are
-                                  // now stored in the redundant_joint_discretization_ member
+  // The next two variables still exists for backwards compatibility
+  // with previously generated custom ik solvers like IKFast
+  // Replace tip_frame_ -> tip_frames_[0], search_discretization_ -> redundant_joint_discretization_
+  MOVEIT_DEPRECATED std::string tip_frame_;
+  MOVEIT_DEPRECATED double search_discretization_;
```
Comment (davetcoleman) [inline moveit_core/kinematics_base/include/moveit/kinematics_base/kinematics_base.h]: i didn't know you could MOVEIT_DEPRECATED a variable. how does this change the behavior of anything?

File: moveit_core/kinematics_base/include/moveit/kinematics_base/kinematics_base.h [hunk 6]
```diff
@@ -569,31 +578,22 @@ class KinematicsBase
...
   std::string group_name_;
   std::string base_frame_;
   std::vector<std::string> tip_frames_;
-  std::string tip_frame_;  // DEPRECATED - this variable only still exists for backwards compatibility with
-                           // previously generated custom ik solvers like IKFast
 
-  double search_discretization_;  // DEPRECATED - this variable only still exists for backwards compatibility
-                                  // with previous implementations.  Discretization values for each joint are
-                                  // now stored in the redundant_joint_discretization_ member
+  // The next two variables still exists for backwards compatibility
```
Comment (davetcoleman) [inline moveit_core/kinematics_base/include/moveit/kinematics_base/kinematics_base.h]: @mlautman any thoughts on these variables w.r.t. IkFast?

File: moveit_core/kinematics_base/include/moveit/kinematics_base/kinematics_base.h [hunk 7]
```diff
@@ -647,6 +647,10 @@ class KinematicsBase
     return false;
   }
 
+  void storeValues(const moveit::core::RobotModel& robot_model, const std::string& group_name,
```
Comment (davetcoleman) [inline moveit_core/kinematics_base/include/moveit/kinematics_base/kinematics_base.h]: add comment - what does this function do? what is it storing where?

File: moveit_core/kinematics_base/src/kinematics_base.cpp [hunk 8]
```diff
@@ -37,42 +37,89 @@
...
+
 namespace kinematics
 {
 const double KinematicsBase::DEFAULT_SEARCH_DISCRETIZATION = 0.1;
 const double KinematicsBase::DEFAULT_TIMEOUT = 1.0;
 
-void KinematicsBase::setValues(const std::string& robot_description, const std::string& group_name,
-                               const std::string& base_frame, const std::string& tip_frame,
-                               double search_discretization)
+static void no_deleter(const moveit::core::RobotModel*)
```
Comment (davetcoleman) [inline moveit_core/kinematics_base/src/kinematics_base.cpp]: why is this function using underscore format?

File: moveit_core/kinematics_base/src/kinematics_base.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (davetcoleman) [reply moveit_core/kinematics_base/src/kinematics_base.cpp]: this is indeed odd and warrants a lot of inline comments as to why this design decision was made. in particular, i'd like to hear your response to @simonschmeisser 's question about raw pointers - why not use those instead?

File: moveit_core/kinematics_base/src/kinematics_base.cpp [hunk 9]
```diff
@@ -37,42 +37,89 @@
...
+
+  ROS_ERROR_NAMED(LOGNAME, "This solver does not support multiple tip frames");
+  return false;
+}
+
+bool KinematicsBase::initialize(const moveit::core::RobotModel& robot_model, const std::string& group_name,
+                                const std::string& base_frame, const std::vector<std::string>& tip_frames,
+                                double search_discretization)
+{
+  ROS_WARN_NAMED("kinematics_base", "IK plugin for group '%s' relies on deprecated API. "
```
Comment (davetcoleman) [inline moveit_core/kinematics_base/src/kinematics_base.cpp]: use LOGNAME

File: moveit_kinematics/cached_ik_kinematics_plugin/include/moveit/cached_ik_kinematics_plugin/cached_ik_kinematics_plugin.h [hunk 10]
```diff
@@ -192,13 +221,23 @@ class CachedIKKinematicsPlugin : public KinematicsPlugin
 
   // virtual methods that need to be wrapped:
 
+  bool initialize(const moveit::core::RobotModel& robot_model, const std::string& group_name,
+                  const std::string& base_frame, const std::vector<std::string>& tip_frames,
+                  double search_discretization) override
+  {
+    return initImpl(robot_model, group_name, base_frame, tip_frames, search_discretization);
```
Comment (davetcoleman) [inline moveit_kinematics/cached_ik_kinematics_plugin/include/moveit/cached_ik_kinematics_plugin/cached_ik_kinematics_plugin.h]: rename to: initializeImpl

File: MIGRATION.md (hunk 3 above)
Comment (rhaschke) [reply MIGRATION.md]: Usually, when my PRs comprise several individual commits, they all have a distinct purpose and can/should be reviewed one-by-one. I don't think that it makes much sense to split these related commits into several PRs.

File: moveit_core/kinematics_base/CMakeLists.txt (hunk 4 above)
Comment (rhaschke) [reply moveit_core/kinematics_base/CMakeLists.txt]: Need to check.

File: moveit_core/kinematics_base/include/moveit/kinematics_base/kinematics_base.h (hunk 5 above)
Comment (rhaschke) [reply moveit_core/kinematics_base/include/moveit/kinematics_base/kinematics_base.h]: You should not use the variable anymore, otherwise you get a deprecation warning.

File: moveit_core/kinematics_base/include/moveit/kinematics_base/kinematics_base.h (hunk 7 above)
Comment (rhaschke) [reply moveit_core/kinematics_base/include/moveit/kinematics_base/kinematics_base.h]: Will add a comment. It's a replacement for `setValues()`.

**Comment by rhaschke** [comment]
@davetcoleman I addressed your formal comments. Do you have any content-wise objections?
Otherwise I will proceed and adapt the remaining kinematics solvers.

**Comment by simonschmeisser** [comment]
Do we really need to migrate all IK plugins at once? as @davetcoleman said this is already quite big

**Comment by rhaschke** [comment]
> Do we really need to migrate all IK plugins at once?

I would like to get rid of corresponding deprecation warnings. As I said, the PR is split up into several individual commits to facilitate review and of course, this PR should not be squash-merged!

**Comment by davetcoleman** [comment]
PRs like this are very hard to review, you can totally cherry-pick the simple style stuff into one or more separate PRs (e.g. "use LOGNAME", "cleanup KinematicsPluginLoad"). Those changes require very little cognitive load, whereas the magical pointer / reference stuff is tricky to understand when mixed in with a bunch of other files.

**Comment by davetcoleman** [comment]
>  Do you have any content-wise objections? Otherwise I will proceed and adapt the remaining kinematics solvers.

It all seems reasonable enough

File: moveit_core/kinematics_base/CMakeLists.txt (hunk 4 above)
Comment (rhaschke) [reply moveit_core/kinematics_base/CMakeLists.txt]: Note, that I don't disable deprecation warnings per se, but only for building `kinematics_base.cpp`.
Deprecation of the variables is foremost to warn derived classes, which still use those variables.

File: moveit_core/kinematics_base/include/moveit/kinematics_base/kinematics_base.h [hunk 11]
```diff
@@ -570,31 +580,22 @@ class KinematicsBase
...
-       // (if multiple tip frames provided, this variable will be unset)
-    search_discretization_(DEFAULT_SEARCH_DISCRETIZATION)
-    , default_timeout_(DEFAULT_TIMEOUT)
-  {
-    supported_methods_.push_back(DiscretizationMethods::NO_DISCRETIZATION);
-  }
+  KinematicsBase();
 
 protected:
+  moveit::core::RobotModelConstPtr robot_model_;
```
Comment (simonschmeisser) [inline moveit_core/kinematics_base/include/moveit/kinematics_base/kinematics_base.h]: Shouldn't we add a warning about the fake nature of this raw pointer in disguise here?

Comment (rhaschke) [reply moveit_core/kinematics_base/include/moveit/kinematics_base/kinematics_base.h]: The shared_ptr can be used as usual. There are no restrictions.

**Comment by jonbinney** [comment]
I'm getting segmentation faults using trac_ik from debs and compiling melodic-devel branch of moveit from source with this PR merged. That's expected, right?

**Comment by rhaschke** [comment]
No, that's not expected. I tested with IKFast only. Can you provide more details / a moveit_config to reproduce this?

**Comment by rhaschke** [comment]
I found an TracIK setup and can reproduce the error. Will have a look.

**Comment by jonbinney** [comment]
Thanks! I did just test with trac_ik from source, and no segfault. So that's good at least.

**Comment by rhaschke** [comment]
I'm afraid that a rebuild is required as KinematicsBase has ABI-incompatible changes.

File: moveit_core/kinematics_base/include/moveit/kinematics_base/kinematics_base.h (hunk 11 above)
Comment (jonbinney) [inline moveit_core/kinematics_base/include/moveit/kinematics_base/kinematics_base.h]: @rhaschke  does adding this `robot_model_` member variable break ABI compatibility? Woudn't that mean all kinematics plugins will need to be recompiled?

Comment (jonbinney) [reply moveit_core/kinematics_base/include/moveit/kinematics_base/kinematics_base.h]: nvm - just saw your comment to this effect.

**Comment by jonbinney** [comment]
Recompiling source code is easy - but this will mean that all IK packages used from debs cannot be used with source version of moveit until moveit is re-released and everything is re-released. @davetcoleman @130s do we have a timeline on next release?

**Comment by rhaschke** [comment]
I'm planning to re-release as soon as #1096 is merged - after the planned sync on this weekend...

**Comment by jonbinney** [comment]
Thanks @rhaschke . I appreciate your work on this!

**Comment by v4hn** [comment]
> but this will mean that all IK packages used from debs cannot be used with source version of moveit until moveit is re-released and everything is re-released.

Do the plugin maintainers have to re-release themselves or is there a way to force rebuilds of non-broken packages on the buildfarm (I would have thought they rebuild everything either way before syncs, but I never looked into this domain before)

Also, the problem will only really be fixed once the next sync is through and this will take a bit anyway.
Thus I don't see the need to release *immediately*.
Given we just merged major changes to the code base I would prefer to hold of the release at least until December 7th (end of next week) to check for potential problems.

**Comment by jonbinney** [comment]
Good point about the sync - and yes, I meant to say that the problem won't be fixed until moveit is re-leased and everything is _rebuilt_ (typo on my part). I do believe that a moveit release will trigger a buildfarm rebuild of all packages which depend on it.

**Comment by rhaschke** [comment]
> I still consider the nodeleter very ugly.

That's a perfectly valid solution. Not breaking API everywhere could be even seen as beautyful ;-)

> Do the plugin maintainers have to re-release themselves?

No, all downstream packages are automatically rebuild and re-released (if they have modeled their dependencies correctly, of course).

**Comment by v4hn** [comment]
> That's a perfectly valid solution.
> Not breaking API everywhere could be even seen as beautyful ;-)

No. It is practical from a maintainance perspective.
Not beautiful from the perspective of good code.

**Comment by rhaschke** [comment]
Ok. I guess, we have different opinions on this...
